### DISCUSSION TITLE
Corfu is broken in `slime-repl-mode`

### DISCUSSION THREAD

**Discussion by djr7C4** [minad/corfu#472] [Q&A]
When performing completion with `corfu` in `slime-repl-mode`, `<return>` is bound to `slime-repl-return` (the major mode binding) even when the corfu popup is active. This overrides corfu's binding of `RET` to `corfu-insert` which causes the current line at the prompt to be submitted when the user only intended to insert the selected completion.

The solution is to simply bind `corfu-insert` to `<return>` in addition to `RET` (similarly to `corfu-complete` with `TAB` and `<tab>`).

**Comment by minad** [comment]
Generally it is not recommended to bind <return> instead of in addition to RET. Ideally the binding should be removed in Slime.

**Reply by djr7C4** [reply]
I understand that in most cases (for example in major modes) it is better to bind `RET` than `<return>`. However, I don't see the disadvantage of binding `<return>` here. Binding `<return>` in a major mode could lead to other `RET` bindings being shadowed but that doesn't seem like a concern with corfu. There may also be other packages that bind `<return>` instead of `RET` and these will have the same problem. Corfu also already binds `<tab>` in addition to `TAB`.

**Reply by minad** [reply]
The main disadvantage is that customization gets harder if both `<return>` and `RET` are bound redundantly. See for example this [customization](https://github.com/minad/corfu?tab=readme-ov-file#tab-only-completion) where the `RET` binding is modified. Also `<return>` only applies to the GUI version of Emacs, potentially leading to inconsistency. Conversely there is no advantage of binding both here, except working around misconfigurations in major modes and I don't see any advantages of binding `<return>` in a major mode. I've also removed the `<tab>` binding in https://github.com/minad/corfu/commit/977ba4b89a68924d6f4137859a8ce6dd0c6884dc, since it is not needed anymore in newer Org versions. Note that these bindings had been marked for removal with a comment all the time.

**Comment by djr7C4** [comment]
I guess it's a difference in philosophy. I prefer that packages be as compatible as possible with other code even if they don't follow best practices unless there is some significant downside to it. The additional customization burden of adding a second binding seems small to me. Anyway, I will just add those bindings to my own Corfu config. Thank you for Corfu and for taking the time to explain your reasoning.

**Reply by minad** [reply]
> I guess it's a difference in philosophy. 

Not much I think.

> I prefer that packages be as compatible as possible with other code even if they don't follow best practices unless there is some significant downside to it.

I prefer if packages are as compatible as possible, but *this applies to both sides*. I've stated that there are downsides, and I consider them somewhat relevant.

Anyway, my position should be pretty much aligned with [the position of the Emacs maintainers](https://lists.gnu.org/archive/html/bug-gnu-emacs/2024-08/msg00798.html) regarding the redundant key binding problem as far as I've understood. That's why Org mode has also removed the conflicting binding by now.
